addons/account/models/account_move.py

```diff
@@ -3230,21 +3223,64 @@ def _get_new_hash(self, secure_seq_number):
         #build and return the hash
         return self._compute_hash(prev_move.inalterable_hash if prev_move else u'')
 
+    def _hash_moves(self, force_hash=False):
+        self.flush_recordset()  # Ensure that the name is correctly computed before it is used to generate the hash
+        for move in self._get_moves_to_hash(force_hash=force_hash):
+            new_number = move.journal_id.secure_sequence_id.next_by_id()
+            super(AccountMove, move).write({
+                'secure_sequence_number': new_number,
+                'inalterable_hash': move._get_new_hash(new_number),
+            })
+
+    def _get_moves_to_hash(self, force_hash=False):
+        """
+        For each journal in the batch of moves, check if the last move in the batch
+        needs to be hashed. If it does, return all the moves that need to be hashed
+        by checking all previous moves (on which the current move depends) in the chain.
+        """
+        res = self.env['account.move']
+        for journal, moves in self.grouped('journal_id').items():
+            last_move = moves.sorted(lambda m: (m.name, m.id))[-1:]
+            if not (
+                last_move and self._is_move_restricted(last_move, force_hash=force_hash)
+            ):
+                return self.env['account.move']
+            domain = self._get_move_hash_domain([
+                ('inalterable_hash', '=', False),
+                ('journal_id', '=', journal.id),
+                ('date', '>', self.company_id._get_user_fiscal_lock_date()),
```

Comment:
It can't depend on the company of the move since moves in the same journal can come from multiple companies
```suggestion
                ('date', '>', journal.company_id._get_user_fiscal_lock_date()),
```

But at the same time it needs to take the lock date of the company of the move into account because that company requires the move to be locked :exploding_head: 

Not sure if we need to apply the suggestion :thinking: